sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala

```diff
@@ -1299,8 +1299,9 @@ class ArrowConvertersSuite extends SharedSparkSession {
          |}
        """.stripMargin
 
-    val a_i = List[Int](1, -1, 2, -2, 2147483647, -2147483648)
-    val b_i = List[Option[Int]](Some(1), None, None, Some(-2), None, Some(-2147483648))
+    // we order values by $"a_i".desc manually as sortBy before coalesce is ignored
```

Comment:
This is the case I mention in the PR description. In this case, `DefaultPartitionCoalescer` does preserve the ordering and the test relied on that even though there is no guarantee it will happen. We could apply the new optimization only if the repartition operation requires a shuffle. That way, we will keep the existing behavior.

Reply:
I think we should apply the optimization only if the repartition requires a shuffle as you suggest. I know that there are users that depend on this behavior.

Reply:
I lean towards that as well.

Reply:
Updated.

Reply:
Could you update the PR description accordingly?